prog/dftb+/lib_type/latpointiter.F90

```diff
@@ -0,0 +1,235 @@
+    end if
+    if (present(reduceByInversion)) then
+      this%tNoInv = reduceByInversion
+    end if
+    if (present(excludeOrigin)) then
+      this%tExcludeOrig = excludeOrigin
+    end if
+
+    this%latVecs(:,:) = latVecs
+    call getRanges(cutoff, invLatVecs, this%posExt, this%negExt, ranges)
+    this%ranges(1,:) = minval(ranges, dim=1)
+    this%ranges(2,:) = maxval(ranges, dim=1)
+    this%curPoint(:) = this%ranges(1,:)
+    this%cutoff2 = cutoff**2
+
+  end subroutine TLatPointIter_init
+
+
+  !> Delivers the next lattice point
+  subroutine TLatPointIter_getNextPoint(this, latticePoint, tFinished)
+
+    !> Instance
+    class(TLatPointIter), intent(inout) :: this
+
+    !> Next lattice point
+    real(dp), intent(out) :: latticePoint(:)
+
+    !> Whether the returned point was the last one.
+    logical, intent(out) :: tFinished
+
+    real(dp) :: rr(3)
+    integer :: curPoint(3)
+
+    curPoint(:) = this%curPoint
+    do
+      tFinished = (curPoint(1) > this%ranges(2, 1))
+      if (tFinished) then
+        exit
+      end if
+      if (this%tNoInv) then
+        if (curPoint(1) < 0) then
+          curPoint(1) = 0
+        end if
+        if (curPoint(2) < 0 .and. curPoint(1) == 0) then
+          curPoint(2) = 0
+        end if
+        if (curPoint(3) < 0 .and. curPoint(1) == 0 .and. curPoint(2) == 0) then
+          curPoint(3) = 0
+        end if
+      end if
+      if (this%tExcludeOrig .and. all(curPoint == 0)) then
+        call increaseInd(this%ranges, curPoint)
+        cycle
+      end if
+      latticePoint(:) = real(curPoint, dp)
+      if (this%tAll) then
+        exit
+      end if
+      rr(:) = latticePoint(1) * this%latVecs(:,1) + latticePoint(2) * this%latVecs(:,2)&
+          & + latticePoint(3) * this%latVecs(:,3)
+      if (sum(rr**2) <= this%cutoff2) then
+        exit
+      end if
+      call increaseInd(this%ranges, curPoint)
+    end do
+
+    if (.not. tFinished) then
+      ! generate image for next call
+      call increaseInd(this%ranges, curPoint)
+    end if
+    this%curPoint(:) = curPoint
+
+  end subroutine TLatPointIter_getNextPoint
+
+
+  !> Returns all lattice points within a cutoff at the same time
+  subroutine TLatPointIter_getAllPoints(this, latticePoints)
+
+    !> Instance
+    class(TLatPointIter), intent(inout) :: this
+
+    !> Lattice points
+    real(dp), allocatable, intent(out) :: latticePoints(:,:)
+
+    real(dp), allocatable :: tmpLatPoints(:,:)
+    integer :: maxLatPoints, iLatPoint
+    logical :: tFinished
+
+    maxLatPoints = product(this%ranges(2,:) - this%ranges(1,:) + 1)
+    allocate(tmpLatPoints(3, maxLatPoints + 1))
+    do iLatPoint = 1, maxLatPoints + 1
+      call this%getNextPoint(tmpLatPoints(:,iLatPoint), tFinished)
+      if (tFinished) then
+        exit
+      end if
+    end do
+    latticePoints = tmpLatPoints(:, 1:iLatPoint - 1)
+
+  end subroutine TLatPointIter_getAllPoints
+
+
+  !> Helper function to increase a tuple of 3 indices by one
```

Comment:
Is this not cycling through permitted values of the tuple within the ranges, instead of just increasing?

Reply:
Yes, it does, sorry for the misleading naming. I've changed it to something (hopefully) more self-explaining.